File: 2021/Day4/Container with most water/Container_with_most_water.py
class Solution:

    def maxArea(self, height: List[int]) -> int:
        """
        Two-pointer approach
        1. one pointer start from the start
        2. one pointer start from the end
        3. calculate the areas
        4. update max area.
        5. return max area once 1st pointer = 2nd pointer
        """
        firstp = 0
        secondp = len(height)-1
        maxarea = 0

        while firstp != secondp:
            currfirst = height[firstp]
            secondfirst = height[secondp]
            hlength = secondp - firstp
            area = hlength*min(currfirst, secondfirst)
            if maxarea < area:
                maxarea = area

            if currfirst >= secondfirst:
                secondp -= 1
            else:
                firstp += 1

        return maxarea

File: 2021/Day4/Container with most water/container_with_most_water.cpp
class Solution //By CyFun
{
public:
    int maxArea(vector<int> &height)
    {
        int lidx = 0, ridx = height.size() - 1, answer = 0;

        while (lidx < ridx)
        {
            int h = min(height[ridx], height[lidx]);
            int square = (ridx - lidx) * h;
            if (answer < square)
                answer = square;

            if (height[ridx] > height[lidx])
                lidx++;
            else
                ridx--;
        }

        return answer;
    }
};